### PR TITLE
Remove deprecated code and OLMKit dependency

### PR DESCRIPTION
**Pull request by Velin92**
This is better reviewed commit by commit.
This is a massive PR about removing legacy code, might split it in multiple PRs, but this is actually the first stable state in which I was able to make everything compile and work

What is done and what is missing:

- [ ]   Remove all sharedHistory code
- [ ]   Remove MX_CRYPTO macros
- [ ]   MXUserTrustLevel can be removed and replaced with a single isVerified boolean, given that we no longer need to distinguish local and crossSigning trust status
- [ ]   NSCoding implementations from MXCrossSigningInfo and others can be removed, as they are no longer persisted on the iOS side, but rather rust side
- [ ]   Remove a number of unnecessary options from MXSDKOptions, such as enableRoomSharedHistoryOnInvite or enableSymmetricBackup
- [x]   A large number of integration tests access a legacyCrypto or legacyCrossSigning property which should be removed. These tests then have to be either refactored or removed entirely
- [x]   Remove MXBackgroundCryptoStore
- [x]   Legacy MXCryptoStore and MXRealmCryptoStore need to remain in place for the purpose of data migration to rust crypto, but only a few methods are required. (NOTE: I actually removed them since we don’t want to allow migration anymore)
- [x]   Remove MXSecretShareManager, and the requestPrivateKeys in MXKeyBackup which is no longer used
- [x]   This will enable the removal of many other associated classes, such as MXLegacySasTransaction, MXMegolmEncryption, MXMegolmDecryption etc
- [x] Remove main ...Legacy... classes such as MXLegacyCrypto, MXLegacyCrossSigning, MXLegacyKeyVerificationManager etc

### CODE REVIEW

File: MatrixSDK/Crypto/Data/MXOlmInboundGroupSession.m
Comment (pixlwave) [inline MatrixSDK/Crypto/Data/MXOlmInboundGroupSession.m:1]: We discussed in a call that we should check whether these `MXOlm…GroupSession` classes are used with Rust crypto or not.

**Review by pixlwave — COMMENTED** [review]
Great stuff 👏

A lot of this goes over my head but we had a chat in a call about it and I've left some summary comments. Definitely needs an approval by someone with more crypto knowledge before merging but looks decent to me.

(Its a shame so many tests are going but they're all clearly based around the legacy crypto stack).

File: MatrixSDK/Crypto/Recovery/MXRecoveryService.m
Comment (Velin92) [reply MatrixSDK/Crypto/Recovery/MXRecoveryService.m:579]: No idea, let's wait for someone from Crypto to confirm this

File: MatrixSDK/IdentityServer/MXIdentityServerRestClient.m
Comment (Velin92) [reply MatrixSDK/IdentityServer/MXIdentityServerRestClient.m]: Okay so I changed the code, honestly I am not sure if what i am doing is the correct way to solve the problem.

**Review by manuroe — COMMENTED** [review]
Thanks for this big cleanup.
I left some questions.

File: MatrixSDK/Crypto/KeyBackup/MXKeyBackupPassword.m
Comment (Velin92) [reply MatrixSDK/Crypto/KeyBackup/MXKeyBackupPassword.m:44]: Yeah this is actually only used in a function that is only used in tests, so I guess we can safely remove it.

File: MatrixSDK/Crypto/KeyBackup/MXRecoveryKey.m
Comment (Velin92) [reply MatrixSDK/Crypto/KeyBackup/MXRecoveryKey.m:75]: okay this value according to [here](https://gitlab.matrix.org/matrix-org/olm/-/blob/master/include/olm/crypto.h#L38) should be 32.

File: MatrixSDK/Crypto/KeyBackup/MXKeyBackupPassword.m
Comment (Velin92) [reply MatrixSDK/Crypto/KeyBackup/MXKeyBackupPassword.m:39]: we can remove it yeah

File: MatrixSDK/Crypto/Recovery/MXRecoveryService.m
Comment (BillCarsonFr) [reply MatrixSDK/Crypto/Recovery/MXRecoveryService.m:579]: Ok, so per my understanding here we are trying to import secrets from the 4S (cross signing keys and megolm backup keys).
- For cross-signing keys, it is supposed to call directly olmMachine.importCrossSigningKeys that will error if secret does not match, and will only store the secret in "cache" if it does.
So by removing the check here it will then change a bit the behavior as the following call to importCrossSigningKeys could then error.

- For backup it is a bit different as all the logic is not in the rust crypto sdk, but still in the native crate. So it appears to me that now **you are not saving the backup decryption key in cache** so the recovery from backup will fail now. I would expect some test to fail here. I would expect the code here to check the validity of the key (without trying to import keys as it would be too long), then to store the secret if valid. Eventually a restore from background could be started

Also looks like you are adding all secrets to `invalidSecrets`, won't this impact/break the UI?

**Review by BillCarsonFr — CHANGES_REQUESTED** [review]
I suspect that some edgecases around recovery are broken now, maybe even happy path for backup?

File: MatrixSDK/Crypto/Recovery/MXRecoveryService.m
Comment (Velin92) [reply MatrixSDK/Crypto/Recovery/MXRecoveryService.m:579]: okay by investigating further, the problem is that this required `- (BOOL)checkSecret:(NSString*)secret withSecretId:(NSString*)secret` that I have removed, and that I could also restore, but relies also on the `MXCrossSigningTools ` object specifically the `isSecretValid:secret forPublicKeys:crossSigningInfo.masterKeys.keys` function, which is heavily using OLMKit.

**Review by BillCarsonFr — APPROVED** [review]
LGTM

**Pull request merged**
